packages/golden-layout/src/utils/ReactComponentHandler.tsx

```diff
@@ -29,14 +29,44 @@ export default class ReactComponentHandler {
     this._container.on('destroy', this._destroy, this);
   }
 
+  /**
+   * Gets the unique key to use for the react component
+   * @returns Unique key for the component
+   */
+  _key(): string {
+    const id = this._container._config.id;
+    if (!id) {
+      throw new Error('Cannot mount panel without id');
+    }
+
+    // If addId is called multiple times, an element can have multiple IDs in golden-layout
+    // We don't use it, but changing the type requires many changes and a separate PR
+    if (Array.isArray(id)) {
+      return id.join(',');
+    }
+
+    return id;
+  }
+
   /**
    * Creates the react class and component and hydrates it with
    * the initial state - if one is present
    *
    * By default, react's getInitialState will be used
+   *
+   * Creates a portal so the non-react golden-layout code still works,
+   * but also allows us to mount the React components in the app's tree
+   * instead of separate sibling root trees
    */
   _render() {
-    ReactDOM.render(this._getReactComponent(), this._container.getElement()[0]);
+    this._container.layoutManager.addReactChild(
+      this._key(),
```

Comment:
Slight optimization - do `const key = this._key()` before addReactChild, since you use it twice here.